pages/form/index.tsx

```diff
@@ -0,0 +1,203 @@
+import React, { FormEvent, useState } from "react";
+import axios from "axios";
+import InputField from "../../components/registration-form/InputField";
+
+/**
+ * Types
+ */
+
+export type FieldState = {
+  value: string;
+  error: string | undefined;
+};
+
+export type FormState = {
+  [fieldName: string]: FieldState;
+};
+
+type FieldDescription = {
+  type: "text" | "password";
+  label: string;
+  placeholder?: string;
+};
+
+/**
+ * Constants
+ */
+
+const fields: { [key: string]: FieldDescription } = {
+  email: {
+    type: "text",
+    label: "Email",
+    placeholder: "email@example.com",
+  },
+  firstName: {
+    label: "First Name",
+    type: "text",
+    placeholder: "Enter your First Name here",
+  },
+  lastName: {
+    label: "Last Name",
+    type: "text",
+    placeholder: "Enter your Last Name here",
+  },
+  password: {
+    label: "Password",
+    type: "password",
+    placeholder: "Enter your Password here",
+  },
+  passwordConfirm: {
+    label: "Confirm Password",
+    type: "password",
+    placeholder: "Confirm your Password",
+  },
+};
+
+const initialFieldState: FieldState = {
+  value: "",
+  error: undefined,
+};
+
+const initialFormState: FormState = {
+  email: initialFieldState,
+  firstName: initialFieldState,
+  lastName: initialFieldState,
+  password: initialFieldState,
+  passwordConfirm: initialFieldState,
+};
+
+/**
+ * Validations
+ */
+
+const validateEmail = (email: string) => {
+  const isValid = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email);
+  return !email
+    ? "Email is required."
+    : !isValid
+    ? "Invalid email address."
+    : undefined;
+};
+
+const validateFirstName = (firstName: string) => {
+  return !firstName.trim() ? "First Name is required." : undefined;
+};
+
+const validateLastName = (lastName: string) => {
+  return !lastName.trim() ? "Last Name is required." : undefined;
+};
+
+const validatePassword = (password: string) => {
+  return !password
+    ? "Password is required."
+    : password.length < 8
+    ? "Password must be at least length of 8"
+    : undefined;
+};
+
+const validatePasswordConfirm = (passwordConfirm: string, state: FormState) => {
+  return !passwordConfirm
+    ? "Confirm Password is required."
+    : state["password"].value !== passwordConfirm
+    ? "Passwords do not match."
+    : undefined;
+};
+
+const fieldsValidations = {
+  email: validateEmail,
+  firstName: validateFirstName,
+  lastName: validateLastName,
+  password: validatePassword,
+  passwordConfirm: validatePasswordConfirm,
+};
+
+/**
+ * Main component
+ */
+
+const RegisterForm = () => {
+  const [state, setState] = useState(initialFormState);
```

Comment:
This is where you would call `useForm`, and then your `FieldDescription` would be a prop to `RegisterForm`